Add support for source acc. in Client#send_payment

jaspervanbrian opened the pull request:
This allows Stellar::Client#send_payment to accept a
source account option.
Generate a test case for sending native assets to an
account through a channel account.

jaspervanbrian committed:
Add support for source acc. in Client#send_payment

This allows Stellar::Client#send_payment to accept a
source account option.
Generate a test case for sending native assets to an
account through a channel account.

ramontayag commented on lib/stellar/client.rb:
```diff
@@ -111,14 +111,25 @@ def send_payment(options={})
+      payment = Stellar::Transaction.payment(payment_details)
+
+      if source_account
+        envelope_base64 = payment.to_envelope(from.keypair, source_account.keypair).to_xdr(:base64)
```
How about shortening this to:

```ruby
to_envelope_args = [from.keypair]
to_envelope_args << source_account.keypair if !source_account.nil?

envelope_base64 = payment.to_envelope(*to_envelope_args).to_xdr(:base64)
```

jaspervanbrian replied:
oh forgot about the splat operator

ramontayag commented on spec/lib/stellar/client_spec.rb:
```diff
@@ -136,6 +137,37 @@
+          b["asset_type"] == "native"
+        end
+        expect(native_asset_balance_info["balance"].to_f).to eq 250.0
+      end
```
You should also test that the source account was the one used. This tests that the amount was sent out, but I think this test will still pass if you comment out `source_account: source` above.

jaspervanbrian committed:
Test the source account was the one used in operation.

Refactor Stellar::Client#send_payment to have more
concise code.

ramontayag commented on lib/stellar/client.rb:
```diff
@@ -111,14 +111,24 @@ def send_payment(options={})
-      })
+      }
+
+      if options[:source_account]
```
You can also shorten this:

```ruby
if source_account = options[:source_account]
  payment_details[:source_account] = source_account.keypair
end

ramontayag commented on spec/lib/stellar/client_spec.rb:
```diff
@@ -136,6 +137,52 @@
+                                          ._attributes
+                                          .instance_variable_get(:@collection)["from"]
+
+        source_account_info = client.account_info(source)
```
I don't think you need to fetch the source account into separately, since we already know that (you generated the account above).

Try:

```ruby
expect(operation_from_address).to eq channel_account.address
```

jaspervanbrian replied:
oh thanks

jaspervanbrian committed:
Refactor client.rb and client_spec.rb

Fix `:hash` to `"hash"` in client_spec.rb on line 164.
Get address of source through `source.address` rather
than fetching through network again.

jaspervanbrian committed:
Merge branch 'master' into add-source-account-option
